lib/pages/class_feedback/service/feedback_provider.dart

```diff
@@ -0,0 +1,158 @@
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+      );
+    } else {
+      return FeedbackQuestion(
+        category: json['category'],
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+      );
+    }
+  }
+}
+
+class FeedbackProvider with ChangeNotifier {
+  Future<bool> addResponse(FeedbackQuestionAnswer response) async {
+    try {
+      await FirebaseFirestore.instance
+          .collection('forms')
+          .doc('class_feedback_answers')
+          .collection(response.questionNumber)
+          .add(response.toData());
+      return true;
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return false;
+    }
+  }
+
+  Future<Map<String, FeedbackQuestion>> fetchQuestions() async {
+    try {
+      final DocumentSnapshot documentSnapshot = await FirebaseFirestore.instance
+          .collection('forms')
+          .doc('class_feedback_questions')
+          .get();
+      final Map<String, dynamic> data = documentSnapshot['questions'];
+      final Map<String, FeedbackQuestion> questions = {};
+      for (final value in data.values) {
+        final key = data.keys.firstWhere((element) => data[element] == value);
+        questions[key] = FeedbackQuestionExtension.fromJSON(value, key);
+      }
+      return questions;
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return null;
+    }
+  }
+
+  Future<Map<String, dynamic>> fetchCategories() async {
+    try {
+      final DocumentSnapshot documentSnapshot = await FirebaseFirestore.instance
+          .collection('forms')
+          .doc('class_feedback_questions')
+          .get();
+      final Map<String, dynamic> data = documentSnapshot['categories'];
+      return data;
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return null;
+    }
+  }
+
+  Future<bool> setUserClassFeedback(String className, String uid) async {
+    try {
+      final DocumentReference ref =
+          FirebaseFirestore.instance.collection('users').doc(uid);
+      await ref.set({
+        'classesFeedback': {className: true}
+      }, SetOptions(merge: true));
+      notifyListeners();
+      return true;
+    } catch (e) {
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return false;
+    }
+  }
+
+  Future<bool> checkProvidedClassFeedback(String className, String uid) async {
+    try {
+      final DocumentSnapshot snap =
+          await FirebaseFirestore.instance.collection('users').doc(uid).get();
+      if (snap.data()['classesFeedback'] != null &&
+          snap.data()['classesFeedback'][className] == true) {
+        return true;
+      }
+      return false;
+    } catch (e) {
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return false;
+    }
+  }
+
+  Future<Map<String, dynamic>> getProvidedFeedbackClasses(String uid) async {
```

Comment:
Again, avoid using dynamic unless you really don't know what the thing is. Here you should know it's a map of strings, no?